### ISSUE TITLE
Setting static outgoing headers causes an InvalidOperationException when sending in parallel

### ISSUE THREAD

**Issue by hmemcpy** [Particular/NServiceBus#3430]
Recently, we released a hotfix for #3091 that fixed a race condition in RijndaelEncryptionService, however this introduced another bug in versions 4.7.10+ and 3.3 - an `InvalidOperationException: Collection was modified; enumeration operation may not execute` is thrown when the `UnicastBus.OutgoingHeaders` is iterated, while another thread sets the outgoing headers.

While sending in parallel is not a thread-safe scenario per our documentation, this does not happen in v5 since a similar issue was fixed in #1976 by making the Dictionary backing the `OutgoingHeaders` property into a ConcurrentDictionary.

Currently, this issue causes PerformanceTests under versions 4.7.x and 3.3 to [fail with this error](http://builds.particular.net/viewLog.html?buildId=143555&tab=buildResultsDiv&buildTypeId=ProductionStagingTests_PerformanceTests). Is there a reason not to apply the same fix to 4.7 (i.e. making the dictionary into a ConcurrentDictionary)?

@johnsimons @ramonsmits @SimonCropp, your input please?

**Comment by johnsimons** [comment]
> Is there a reason not to apply the same fix to 4.7 (i.e. making the dictionary into a ConcurrentDictionary)?

Nope

**Comment by SimonCropp** [comment]
@hmemcpy the `UnicastBus.OutgoingHeaders` is designed to only be used at configuration/startup time. it was never intended to be changed at runtime. hence it was not designed with concurrency in mind. arguable the ConcurrentDictionary change actually enabled bad practice and potential bugs. this has been fixed in v6 to be a config only option. 

so arguable the bug here is that something is using OutgoingHeaders at runtime in v4 and v3.

**Comment by andreasohlund** [comment]
>so arguable the bug here is that something is using OutgoingHeaders at runtime in v4 and v3.

@ramonsmits the encryption fix is using the outgoing headers? (can we use the pipeline executor instead in some way?)

**Comment by timbussmann** [comment]
so can we close this one as "won't fix"?

**Comment by hmemcpy** [comment]
Well, performance tests are now broken in v4.7. Before the fix, it was possible to workaround the encryption race condition bug by setting the headers in configuration, but now I don't see a way to workaround iterating the collection anymore, other than using ConcurrentDictionary (like in v5).

I agree completely with @SimonCropp, but I'd also like a solution for the problem at hand.

**Comment by danielmarbach** [comment]
Removed @ramonsmits from being assigned to this.

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers Like @timbussmann already proposed I think we should close this as won't fix.

**Comment by andreasohlund** [comment]
:+1:
